lib/filter/tests/test_filters_fop_cmp.c

```diff
@@ -311,11 +311,33 @@ Test(filter, test_type_aware_comparison_strings_are_compared_as_strings_if_types
   cr_assert(evaluate("'$strvalue' < 'zabkasa')"));
 }
 
-Test(filter, test_type_aware_comparison_null_equals_to_null_and_converts_to_zero)
+Test(filter, test_type_aware_comparison_null_equals_to_null_and_does_not_equal_to_anything_else)
 {
   cfg_set_version_without_validation(configuration, VERSION_VALUE_4_0);
+  cr_assert(evaluate("null('') == null('')"));
+  cr_assert_not(evaluate("null('') != null('')"));
   cr_assert(evaluate("'$nullvalue' == null('')"));
-  cr_assert(evaluate("'$nullvalue' == 0"));
+  cr_assert_not(evaluate("'$nullvalue' != null('')"));
+  cr_assert(evaluate("string('') != null('')"));
+  cr_assert(evaluate("int64('0') != null('')"));
+  cr_assert(evaluate("double('0.0') != null('')"));
```

Comment:
This could work if the patch supporint numbers in templates with type-hint applied from #4025 :
```
diff --git a/lib/cfg-grammar.y b/lib/cfg-grammar.y
index 8a80adb6c..30f912587 100644
--- a/lib/cfg-grammar.y
+++ b/lib/cfg-grammar.y
@@ -804,7 +804,7 @@ template_content_inner
           CHECK_ERROR_GERROR(log_template_compile(last_template, $1, &error), @1, error, "Error compiling template");
           free($1);
         }
-        | LL_IDENTIFIER '(' string ')'
+        | LL_IDENTIFIER '(' string_or_number ')'
         {
           GError *error = NULL;
```

Reply:
This hunk is already in a followup branch (the one to support === and !== operators).

Reply:
This is optional, not important. :) Let's keep the string type in the test as You've fixed it:

`cr_assert(evaluate("double('0.0') != null('')"));`

Reply:
I don't understand exactly what you mean here, sorry. Can you elaborate?

Reply:
It's probably nothing, but this is what I meant:
The grammar doesn't support numbers in a template after a type-hint, so you've changed the template in the unit tests to fix the failing unit tests.

```diff
-  cr_assert(evaluate("int64(0) != null('')"));
-  cr_assert(evaluate("double(0.0) != null('')"));
+  cr_assert(evaluate("int64('0') != null('')"));
+  cr_assert(evaluate("double('0.0') != null('')"));
```

I've seen that the template's grammar rule is changed in #4025 and numbers will be supported in a type-hinted template, therefore I've suggested using it in this PR too.